### ISSUE TITLE
Carthage 0.15.2 checks out stale commit of branch in local repo

### ISSUE THREAD

**Issue by t0rst** [Carthage/Carthage#1339]
* `carthage version`: 0.15.2
* `xcodebuild -version`: Xcode 7.3.1 Build version 7D1014
* Are you using `--no-build`? YES
* Are you using `--no-use-binaries`? YES
* Are you using `--use-submodules`? NO

**Cartfile**
```
git "file:///Users/me/Dev/OBP/Dev/OBPSDK" "daily"
github "tonymillion/Reachability" "35fdc856300fb6c472de53f6e78a164c790b7202"
github "shpakovski/MASPreferences"
```

**Carthage Output**
```
*** Fetching MASPreferences
*** Fetching UICKeyChainStore
*** Fetching STHTTPRequest
*** Fetching OAuthCore
*** Checking out MASPreferences at "1.1.4"
*** Checking out OAuthCore at "0.0.2"
*** Checking out Reachability at "35fdc856300fb6c472de53f6e78a164c790b7202"
*** Checking out STHTTPRequest at "1.1.2"
*** Checking out UICKeyChainStore at "v2.1.0"
*** Checking out OBPSDK at "c84c4b9d89fac197235f08be3324823318cceca7"
```

**Full Terminal Listing**
```
MacBook-Pro:TimeBalance me$ carthage update --no-build --no-use-binaries
*** Fetching MASPreferences
*** Fetching UICKeyChainStore
*** Fetching STHTTPRequest
*** Fetching OAuthCore
*** Checking out MASPreferences at "1.1.4"
*** Checking out OAuthCore at "0.0.2"
*** Checking out Reachability at "35fdc856300fb6c472de53f6e78a164c790b7202"
*** Checking out STHTTPRequest at "1.1.2"
*** Checking out UICKeyChainStore at "v2.1.0"
*** Checking out OBPSDK at "c84c4b9d89fac197235f08be3324823318cceca7"
MacBook-Pro:TimeBalance me$ cat cartfile
git "file:///Users/me/Dev/OBP/Dev/OBPSDK" "daily"
github "tonymillion/Reachability" "35fdc856300fb6c472de53f6e78a164c790b7202" # commit at which framework building project was added; latest version tag is before this; remove when there is a newer version tag
github "shpakovski/MASPreferences"
MacBook-Pro:TimeBalance me$ cd /Users/me/Dev/OBP/Dev/OBPSDK
MacBook-Pro:OBPSDK me$ git branch -av
  daily                  371f8a0 160608_2121
  master                 bb55d0a OBPKit 1.0.0
* ptr                    371f8a0 160608_2121
  remotes/dropbox/daily  ba0bb14 160607_2128
  remotes/dropbox/master bb55d0a OBPKit 1.0.0
  remotes/origin/master  bb55d0a OBPKit 1.0.0
MacBook-Pro:OBPSDK me$ git reflog
371f8a0 HEAD@{0}: commit: 160608_2121
c84c4b9 HEAD@{1}: commit: 160608_18355
23f288f HEAD@{2}: commit: 160608_1839
ed1a5a8 HEAD@{3}: commit: 160608_1839
ba0bb14 HEAD@{4}: reset: moving to daily
MacBook-Pro:OBPSDK me$ 
MacBook-Pro:OBPSDK me$ carthage version
0.15.2
```

**Problem Description**

Cartfile requests latest commit of branch 'daily' in local repo OBPSDK
Latest commit on local repo OBPSDK branch daily is `371f8a0`, but `carthage update --no-build --no-use-binaries` checks out `c84c4b9`. This is a previous value that is no longer on branch daily. My branches 'ptr' and 'daily' used to point at `c84c4b9`, but then I used `git commit --amend` to include a small omission, and the resulting new commit was `371f8a0`; immediately after this I used `git branch -f daily` to point daily at the new head (`371f8a0`).

I returned to other projects, executed `carthage update --no-build --no-use-binaries ` to use the corrected OBPSDK source, but did not get the updated source. Couldn't find a way of forcing a refresh - deleting cartfile.resolved and the Carthage directory did not help. So its quite persistent. Have seen it before, probably around carthage 0.14.0.

In a nutshell: `git branch -f branch_name new_commit` in a local repo seems to fly underneath carthage's radar.

**Comment by mdiep** [comment]
Can you upgrade to 0.16.2 and try again? We've had some improvements here, and I think this may be fixed.

**Comment by ikesyo** [comment]
Yes, this should be fixed in [0.16](https://github.com/Carthage/Carthage/releases/tag/0.16).

> `update` will fetch when dependencies are pinned to a branch (#1202).

**Comment by t0rst** [comment]
Yep, fixed in 0.16.2. Cheers!
